File: ticket_system.py
version = "1.0.1"
class Ticket: # creating a ticket class to be used in an array of tickets
	num = 0
	name = ""
	staff_id = -1
	email = ""
	description = ""
	response = "Not Yet Provided"
	status = "Open"
	resolved = False

	def __init__(self, num, name, staff_id, email, description): # constructor that is run when the object is created
		self.num = num
		self.name = name
		self.staff_id = staff_id
		self.email = email
		self.description = description
		if(description == "Password change"):
			self.password_change()
			print("Password changed")

	def submit_ticket(self, name, staff_id, email, description): # class method to submit ticket
		self.num = num
		self.name = name
		self.staff_id = staff_id
		self.email = email
		self.description = description

	def resolve_ticket(self, response):
		self.response = response
		self.status = "Closed"
		self.resolved = True

	def is_resolved(self):
		return self.resolved 

	def password_change(self):
		# self.resolve_ticket(self.staff_id[:2] + self.name[:3])
		name = self.name.split()
		password = "".join([self.staff_id[:2], name[0][:3]])
		self.resolve_ticket(password)
		print(password)
		return 

	def display(self):
		print("Ticket number: {}\nTicket Creator: {}\nStaff ID: {}\nEmail Address: {}\nDescription: {}\nResponse: {}\nTicket Status: {}".format(self.num, self.name, self.staff_id, self.email, self.description, self.response, self.status))

def main(): # Main function
	print("Ticketing system version:", version)
	ticket_list = []
	ticket_num = 2000
	while(1):
		command = input("What would you like to do (m for help): ")
		if(command == 'm'):
			print("\nn  create new ticket\nd  display statistics\np  print tickets\nr  respond to ticket\nq  quit\n")
			continue
		if(command == 'n'):
			staff_id = input("\nStaff ID: "	)
			name = input("Name: ")
			email = input("Email: ")
			description = input("Description: ")
			if(len(staff_id) < 2):
				print("Staff ID too short\n")
				continue
			if(len(name) < 3):
				print("Name too short\n")
				continue
			ticket_list.append(Ticket(ticket_num, name, staff_id, email, description))
			print("\n")
			ticket_num += 1
		if(command == 'd'):
			print("\n")
			resolved = 0
			for i in ticket_list:
				if(i.is_resolved()):
					resolved += 1
			print("Tickets Created: {}\nTickets Resolved: {}\nTickets To Solve: {}\n".format(len(ticket_list), resolved, len(ticket_list) - resolved))	
		if(command == 'p'):
			print("\n")
			for i in ticket_list:
				i.display()
				print("\n")
		if(command == 'r'):
			print("\n")
			for i in ticket_list:
				i.display()
				print("\n")
			choice = input("\nWhich ticket would you like to respond to: ")
			for i in ticket_list:
				if(str(i.num) == choice and bool(i.resolved) == False):
					i.display()
					print("\n")
					response = input("\nWhat is your response to this ticket: ")
					if(len(response) < 1):
						break
					i.resolve_ticket(response)
			print("\n")
		if(command == 'q'):
			print("Exiting...")
			quit()



if __name__ == "__main__": # Essentially if the code is being run by itself run main. This is useful for when you want to make a code that can be run by itself or be used in part of another script
	main()

File: ticket_system.cpp
#include<iostream>
#include<string>
using namespace std;

class Ticket{
	private:
		int num;
		string name;
		string staff_id = "Unset";
		string email;
		string description;
		string response = "Not Yet Provided";
		string status = "Open";
		bool resolved = false;

		void password_change(){
			resolve_ticket(staff_id.substr(2) + name.substr(3));
		}

	public:
		void create(int num_i, string name_i, string staff_id_i, string email_i, string description_i){
			num = num_i;
			name = name_i;
			staff_id = staff_id_i;
			email = email_i;
			description = description_i;
		}

		bool is_resolved(){
			return resolved;
		}

		void resolve_ticket(string resolution){
			response = resolution;
			status = "Closed";
			resolved = true;
		}

		void display_ticket(){
			cout << "Ticket number: " << num << "\nTicket Creator: " << name << "\nStaff ID: " << staff_id << "\nEmail Address: " << email << "\nDescription: " << description << "\nResponse: " << response << "\nTicket Status: " << status << endl;
		}
};

int main(){
	Ticket ticket_list[200];
	int ticket_num = 0;
	int ticket_num_start = 2000;
	string command;
	while(1){
		cout << "What would you like to do (m for help): ";
		cin >> command;
		if(command == "m"){
			cout << "n  create new ticket\nd  display statistics\np  print tickets\nr  respond to ticket\nq  quit" << endl;
			continue;
		}
		if(command == "n"){
			if(ticket_num == sizeof(ticket_list)/sizeof(*ticket_list)){
				cout << "No space to create new ticket";
				continue;
			}
			string staff_id_t;
			string name_t;
			string email_t;
			string description_t;
			cout << "Staff ID: ";
			cin >> staff_id_t;
			cout << "Name: ";
			cin >> name_t;
			cout << "Email: ";
			cin >> email_t;
			cout << "Description: ";
			cin >> description_t;
			if(staff_id_t.length() < 2){
				cout << "Staff ID too short" << endl;
				continue;
			}
			if(name_t.length() < 2){
				cout << "Name too short" << endl;
				continue;
			}
			ticket_list[ticket_num].create(ticket_num+ticket_num_start, name_t, staff_id_t, email_t, description_t);
			ticket_num++;
			continue;
		}
		if(command == "d"){
			int resolved = 0;
			for(int i = 0; i < ticket_num; i++){
				ticket_list[i].display_ticket();
				if(ticket_list[i].is_resolved()){
					resolved++;
				}
			cout << "Tickets Created: " << ticket_num << "\nTickets Resolved: " << resolved << "\nTickets To Solve: " << ticket_num - resolved << endl;
			}
			continue;
		}
		if(command == "p"){
			for(int i = 0; i < ticket_num; i++){
				ticket_list[i].display_ticket();
			}
		}
		if(command == "r"){
			for(int i = 0; i < ticket_num; i++){
				ticket_list[i].display_ticket();
			}
			unsigned int num = 0;
			cout << "Select Ticket by number: ";
			cin >> num;	
			num -= ticket_num_start;
			if(num > ticket_num){
				continue;
			}
			ticket_list[num].display_ticket();
			string resolution;
			cin >> resolution;
			if(resolution.length() < 1){
				continue;
			}
			ticket_list[num].resolve_ticket(resolution);
			continue;
		}
		if(command == "q"){
			cout << "Exiting..." << endl;
			return 0;
		}
	}
	return 0;
}
